[jit] fix handling of function attributes.

suo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #28395 [jit] fix @property
* **#28569 [jit] fix handling of function attributes.**

Previously, the inclusion of function attributes would "poison" a
ConcreteModuleType, because we did not have a way of checking whether
they are actually the same function. This PR uses the Python function
object to perform that check. This improves our ability to reuse JIT
types between modules.

Also this PR fixes a bug where we weren't properly adding modules as
attributes when converting from ConcreteType -> JIT type (we were adding
them after the fact--another reason to switch from using `register_x` to
`set_x` during module construction, which is on my to-do list after
this).

Fixes https://github.com/pytorch/pytorch/issues/28559

Pull Request resolved: https://github.com/pytorch/pytorch/pull/28569

Differential Revision: [D18111331](https://our.internmc.facebook.com/intern/diff/D18111331)

suo committed:
[jit] fix handling of function attributes.

Previously, the inclusion of function attributes would "poison" a
ConcreteModuleType, because we did not have a way of checking whether
they are actually the same function. This PR uses the Python function
object to perform that check. This improves our ability to reuse JIT
types between modules.

Also this PR fixes a bug where we weren't properly adding modules as
attributes when converting from ConcreteType -> JIT type (we were adding
them after the fact--another reason to switch from using `register_x` to
`set_x` during module construction, which is on my to-do list after
this).

Fixes https://github.com/pytorch/pytorch/issues/28559

suo committed:
Update on "[jit] fix handling of function attributes."


Previously, the inclusion of function attributes would "poison" a
ConcreteModuleType, because we did not have a way of checking whether
they are actually the same function. This PR uses the Python function
object to perform that check. This improves our ability to reuse JIT
types between modules.

Also this PR fixes a bug where we weren't properly adding modules as
attributes when converting from ConcreteType -> JIT type (we were adding
them after the fact--another reason to switch from using `register_x` to
`set_x` during module construction, which is on my to-do list after
this).

Fixes https://github.com/pytorch/pytorch/issues/28559

Differential Revision: [D18111331](https://our.internmc.facebook.com/intern/diff/D18111331)

driazati commented on test/jit/test_type_sharing.py:
```diff
@@ -301,6 +322,26 @@ def forward(self, x):
 
+    def test_python_function_attribute_same(self):
+        """
+        Different functions passed in should lead to different types
```
Different -> same?

driazati commented on torch/csrc/jit/script/concrete_module_type.cpp:
```diff
@@ -36,6 +36,11 @@ ClassTypePtr ConcreteModuleType::createNewTypeFromThis() {
     cls->addAttribute(name, type, isParameter);
   }
 
+  for (const auto& moduleInfo : modules_) {
```
What is this adding to `cls`? Is `modules_` all callables attached to this type or is it just sub-`ScriptModule`s?

suo replied:
it's all the submodules

suo commented on test/jit/test_type_sharing.py:
```diff
@@ -301,6 +322,26 @@ def forward(self, x):
+        fn1_mod = M(fn)
+        fn2_mod = M(fn)
+
+        self.assertSameType(fn1_mod, fn2_mod)
```
that's test_python_function_attribute_different

suo commented on test/jit/test_recursive_script.py:
```diff
@@ -531,6 +531,30 @@ def forward(self, x):
+            def __init__(self):
+                super().__init__()
+                encoder_norm = nn.ReLU()
+                self.encoder = N(encoder_norm)
```
oh my attempt to minify this test case went wrong. I'll add back the function attr

suo committed:
Update on "[jit] fix handling of function attributes."


Previously, the inclusion of function attributes would "poison" a
ConcreteModuleType, because we did not have a way of checking whether
they are actually the same function. This PR uses the Python function
object to perform that check. This improves our ability to reuse JIT
types between modules.

Also this PR fixes a bug where we weren't properly adding modules as
attributes when converting from ConcreteType -> JIT type (we were adding
them after the fact--another reason to switch from using `register_x` to
`set_x` during module construction, which is on my to-do list after
this).

Fixes https://github.com/pytorch/pytorch/issues/28559

Differential Revision: [D18111331](https://our.internmc.facebook.com/intern/diff/D18111331)
